aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -71,24 +72,31 @@ struct AT_CUDA_API CUDAEvent {
     if (!was_recorded_) record(stream);
   }
 
+  // Note: cudaEventRecord must be called on the same device as the STREAM
   void record(const CUDAStream& stream) {
+    at::cuda::CUDAGuard device_index_guard(static_cast<int16_t>(stream.device_index()));
```

Comment:
Not really for this PR, but we should eliminate the need for this static cast at some point. `device_index` and `CUDAGuard` should produce and accept `int16_t`, no problem. And in any case, we should use `DeviceIndex` instead to represent the type.